README.md
# TlsTestGate Azure DevOps Extension

## Overview

Test public or internal endpoints for non-compliant SSL/TLS settings. The extension can be inserted into a build or release pipeline and can serve as a compliance gate. This extension leverages the [SSLyze](https://github.com/nabla-c0d3/sslyze) API. Currently, SSL 2.0/3.0 and TLS 1.0/1.1 cipher suites are considered non-compliant. Additionally, certain TLS 1.2 ciphers that are considered "weak" are also considered non-compliant.

## Valid TLS 1.2 Ciphers

Some cipher suites within TLS 1.2 are considered [weak](https://blog.qualys.com/product-tech/2019/04/22/zombie-poodle-and-goldendoodle-vulnerabilities#). The cipher suites listed below are considered "secure" for now and may be updated in the future:

* TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
* TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
* TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
* TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
* TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
* TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
* TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256

## Usage

There are four inputs to the extension: Base URL, port, DNS server, and an option to fail the build or release:

* "Base URL" should be the DNS domain name or IP that you would like to scan (i.e., github.com or 140.82.113.4)

* "port" is optional and if omitted will default to TCP 443

* Since corporations often use [split-view DNS](https://en.wikipedia.org/wiki/Split-horizon_DNS), "DNS server" in this context is the network viewpoint you want to scan, either internal or external. This is accomplished by specifying a valid DNS server to use for name resolution. The default value for external will use Google (e.g. 8.8.8.8)

* "Fail task" will fail the build or release if non-compliant SSL/TLS settings are identified. The default is to publish the test results whether or not non-compliant settings are identified. Check the box if you want to gate on a failing test.

### Classic Pipeline Example

![TlsTestGate Extension](https://github.com/MetLife/TlsTestGate/blob/master/images/tlstestgateextension.png)

### YAML sample

Below is sample YAML to insert into your build or release pipeline.

```
steps:
- task: JoeGatt.TlsTestGate.custom-build-release-task.TlsTestGate@1
  displayName: 'github.com SSL/TLS Test Gate'
  inputs:
    baseURL: github.com
    port: 443
    dnsserver: 8.8.8.8
    failBuild: true
```

## Results
Vulnerabilities (if any) are automatically published to the build or release pipeline. To view them, simply click on the "Tests" tab. For each vulnerability discovered, a "failed test" will appear in the results.

![TlsTestGate Summary Results](https://github.com/MetLife/TlsTestGate/blob/master/images/resultsummary.png)

![TlsTestGate Summary Detail](https://github.com/MetLife/TlsTestGate/blob/master/images/resultdetail.png)

## Fixing Issues Identified by TlsTestGate

The Mozilla SSL Configuration [Generator](https://ssl-config.mozilla.org/) is an excellent resource to use to securely configure a web server. However, SSL/TLS settings are also often set on load balancers or reverse proxies. Fixing your local web server config may not fix the issue, depending on your network topology.

# Pre-requisites to Develop Locally

For more information on developing Azure DevOps extensions, refer to this [link](https://docs.microsoft.com/en-us/azure/devops/extend/develop/add-build-task?view=azure-devops) from Microsoft.

Development Environment Setup:

1. Create and setup virtual environment for python:
```bash
python3 -m venv foo
cd foo
source bin/activate
git clone https://github.com/metlife/tlstestgate.git
cd tlstestgate
pip install -r requirements.txt
```

2. Environment setup at a high level:

* Download and install node from [here](https://nodejs.org/en/download/).

* Install the typescript compiler: `npm install -g typescript`

* Install the TFS Cross Platform Command Line Interface (tfx-cli): `npm install -g tfx-cli`

* Install node packages:
```bash
cd buildAndReleaseTask
npm init
npm install azure-pipelines-task-lib
npm install azure-pipelines-tool-lib
npm install @types/node --save-dev
npm install @types/q --save-dev
```

3. Compile the extension:
```bash
cd buildAndReleaseTask
tsc
```
    
4. Test the extension locally:
```bash
cd buildAndReleaseTask
export INPUT_BASEURL = github.com
export INPUT_PORT = 443
export INPUT_DNSSERVER = 8.8.8.8
export INPUT_FAILBUILD = false
node tlstestgate.js
```

## Testing the Scanner

To test the TLS/SSL scanning functionality, run:

```bash
cd tests
pytest -v test_tlstestgate.py --cov=buildAndReleaseTask --cov-report=xml
```

## Known Issues and Limitations of the Microsoft hosted Azure Pipeline agent

If you intend to test a private endpoint, it is probable that the Microsoft hosted agents do not have access to your internal network. If you want to test a private endpoint, please use a self-hosted Azure Pipeline agent. For self-hosted agents, Python 3.7 and above is required. Please refer to the links below for your target platform:

* [Linux](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-linux?view=azure-devops)
* [MacOS](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-osx?view=azure-devops)
* [Windows](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-windows?view=azure-devops)

The location of the latest self-hosted agents is [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-windows?view=azure-devops)

## References

[Here](https://www.paraesthesia.com/archive/2020/02/25/tips-for-custom-azure-devops-build-tasks/) are some useful tips for developing tasks for Azure DevOps.


CHANGELOG.md
# Changelog for TlsTestGate

# v1.0.0
- Initial Release

# v1.1.0
- Version bump for sslyze and junitparser
- Added tox for testing scanner functionality on Python 3.7 and 3.8
- Fixed dns resolution error handling
- Added azure-pipelines.yaml to automate tox testing

# v1.2.0
- Rename internal variable (non breaking change)
- Update node packages
- Improved documentation with screen shots
- Imporved python version detection and error handling

# v1.3.0
- Added support for Python 3.9 with upgrade of SSLyze to 4.0.x
- Added support for junitparser 2.0.0 (fixes breaking change)
- Removed hardcoded python package configuration from tlstestgate.ts
- Added tox tests for Python 3.9
- Added pytest to validate junitparser functionality in scanner.py

# v1.4.0
- Created docker development environment in VS Code to work around M1 compatibility issues with SSLyze/NaSSL
- Updated to SSLyze 5.0.0
- Changed minimum Azure Pipeline Agent version to 2.144.0 in support of Node10
- Updated azure-pipelines-task-lib and azure-pipelines-tool-lib to latest version
- Formatted python files with black and linted with pylint
- Fixed test case that purposely failed resolving bbbbbbbbbbbbbbb.com because someone out there is now using it

# v1.0.10
- Align changelog version scheme with Azure DevOps Marketplace
- Removed chai, mocha, tfx-cli, ts-node, and typescript from dev-dependencies and moved install into Dockerfile
- Added eslint to project and pipelines
- Updated SSLyze to 5.0.3 for Python 3.10 support
- Removed is-Ip TypeScript dependency
- Fixed test case that purposely failed resolving bbbbbbbbbbbbbbbbbbbbbbbbb.com because someone out there is now using it

# v1.0.11
- Fix yaml example on Visual Studio marketplace

# v1.0.12
- Updated all npm and python dependencies
- Added dependabot
- Added Microsoft Typescript security linters
- Removed deprecated node api call (url.parse)
- Moved CI from Azure DevOps to GitHub Actions


tests/test_tlstestgate.py
""" TlsTestGate pytest tests

usage: pytest -v test_tlstestgate.py --cov=buildAndReleaseTask

"""
import html
import json
import os

from junitparser import JUnitXml

from buildAndReleaseTask.python.scanner import scan, write_output

DNS_SERVER = "8.8.8.8"

TEST_RESULTS = {
    "Results": [
        {"Version": "TLS_1_1", "Cipher": "TLS_RSA_WITH_AES_256_CBC_SHA"},
        {"Version": "TLS_1_1", "Cipher": "TLS_RSA_WITH_AES_128_CBC_SHA"},
        {"Version": "TLS_1_1", "Cipher": "TLS_RSA_WITH_3DES_EDE_CBC_SHA"},
        {"Version": "TLS_1_1", "Cipher": "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"},
        {"Version": "TLS_1_1", "Cipher": "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"},
        {"Version": "TLS_1_0", "Cipher": "TLS_RSA_WITH_AES_256_CBC_SHA"},
        {"Version": "TLS_1_0", "Cipher": "TLS_RSA_WITH_AES_128_CBC_SHA"},
        {"Version": "TLS_1_0", "Cipher": "TLS_RSA_WITH_3DES_EDE_CBC_SHA"},
        {"Version": "TLS_1_0", "Cipher": "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"},
        {"Version": "TLS_1_0", "Cipher": "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"},
    ]
}


def test_scan_no_violations():
    """Test scan on a host with no violations"""

    results = scan(DNS_SERVER, "api.metlife.com", 443)

    # Check the output to ensure there are no violations
    assert results["Results"] == ["No SSL/TLS Violations found."]


def test_scan_with_violations():
    """Test scan on a host with violations"""

    results = scan(DNS_SERVER, "espn.com", 443)

    # Check the output to ensure there are no violations
    assert results["Results"] != ["No SSL/TLS Violations found."]


def test_dns_name_not_resolved():
    """Test dns name not resolved"""

    results = scan(DNS_SERVER, "joegatt.com", 443)

    # Check the output to ensure the DNS name could not resolve
    assert results["Results"] == ["joegatt.com exists but no A record."]


def test_external_dns_name_not_exist():
    """Test NXDOMAIN"""

    results = scan(DNS_SERVER, "jeogatt.com", 443)

    # Check the output to ensure the DNS name could not resolve
    assert results["Results"] == ["The DNS name jeogatt.com does not exist."]


def test_sslyze_timeout():
    """Test sslyze timeout"""

    bad_dns_domain = "bbbbbbbbbbbbbbbbbbbbbbbbbb.com"
    results = scan(DNS_SERVER, bad_dns_domain, 443)

    # Check the output to ensure the DNS name could not resolve
    assert results["Results"] == [f"The DNS name {bad_dns_domain} does not exist."]


def test_bad_dnsserver():
    """Test bad dns server"""

    results = scan("8.8.7.6", "espn.com", 443)

    # Check the output to ensure the DNS name could not resolve
    assert results["Results"] == ["The DNS operation timed out."]


def test_port_timeout():
    """Test timeout connecting to a port"""

    results = scan(DNS_SERVER, "yahoo.com", 8443)

    # Check the output to ensure the scan could not complete
    assert results["Results"] == ["Error: Could not connect to yahoo.com:8443"]


def test_by_ip_no_violations():
    """Test policy scan on an external ip with no violations"""

    results = scan(DNS_SERVER, "216.163.251.205", 443)

    # Check the output to ensure there are no violations
    assert results["Results"] != ["No SSL/TLS Violations found."]


def test_junit_parser_with_violations():
    """Test writing results to test-output.xml and reading output"""

    # Write the test-output.xml
    write_output("foo", TEST_RESULTS)

    # Load it
    output = os.path.normpath(
        os.path.abspath(os.path.expanduser(os.path.expandvars("test-output.xml")))
    )

    xml = JUnitXml.fromfile(output)
    for suite in xml:
        # Should be one failure
        assert suite.failures == 1

        for case in suite:
            for element in case:
                results = html.unescape(element.message)
                json_results = json.loads(results.replace("'", '"'))
                # Should be 10 results
                assert len(json_results["Results"]) == 10


buildAndReleaseTask/test/getPythonPath.test.ts
import { getPythonPath } from '../getpythonpath';
import 'mocha';
const chai = require('chai'); // eslint-disable-line @typescript-eslint/no-var-requires
const expect = chai.expect;

describe('getPythonPath.ts tests', function() {
  describe('Determine python path on a Microsoft or self-hosted agent', function() {
    it('Should return python path.', async function() {

      // Get the Python path
      expect(function () {getPythonPath}).to.not.throw();

      const pythonPath: string = await getPythonPath();
      console.log('pythonPath: ' + pythonPath);
      expect(pythonPath).to.include('python');

    });
  });
});


buildAndReleaseTask/tlstestgate.ts
/**
 * There are four parts to the extension: Base URL, port, DNS server, and decision
 * @baseURL - url to scan.
 * @port (optional) - port to scan (default is 443)
 * @dnsserver dnsserver to use (default is 8.8.8.8)
 * @failBuild whether or not to fail the task on failed test
 */

import * as tl from 'azure-pipelines-task-lib/task';
import type trm from 'azure-pipelines-task-lib/toolrunner';
import * as net from 'net';
import * as path from 'path';
import { getPythonPath } from './getpythonpath';

async function run(): Promise<void> {
    try {
        tl.setResourcePath(path.join(__dirname, 'task.json'));

        const baseURL: string = <string>tl.getInput('baseURL', true);
        const port: string = <string>tl.getInput('port', true);
        const dnsserver: string = <string>tl.getInput('dnsserver', true);
        const decision: boolean = <boolean>tl.getBoolInput('failBuild', true);

        // Get failTaskOnFailedTests preference
        let testDecision: string;
        if (decision === false) {
            testDecision = 'pass';
        } else {
            testDecision = 'fail';
        }

        // Check targetdomain formatting
        let hostname: string | null = baseURL;
        if (hostname.startsWith('http')) {
            const myURL = new URL(hostname);
            hostname = myURL.hostname;
        }

        // Check port number
        if (Number(port) > 65535) {
            throw new Error(tl.loc('portNotValid'));
        }

        // Validate the DNS server formatting
        if (!net.isIPv4(dnsserver)) {
            throw new Error(tl.loc('dnsserverNotValid'));
        }

        // Get Python 3 path
        const pyPath: string = await getPythonPath();
        console.log('PYTHON PATH: ' + `${pyPath}`);

        try {
            // Install setuptools and wheel
            const pythonsetup: trm.ToolRunner = tl.tool(pyPath);
            pythonsetup.arg('-m');
            pythonsetup.arg('pip');
            pythonsetup.arg('install');
            pythonsetup.arg('--upgrade');
            pythonsetup.arg('pip');
            pythonsetup.arg('setuptools');
            pythonsetup.arg('wheel');
            await pythonsetup.exec();
            tl.setResult(tl.TaskResult.Succeeded, 'python setup was successful.');

        } catch (err: any) { // eslint-disable-line @typescript-eslint/no-explicit-any

            return tl.setResult(tl.TaskResult.Failed, 'python setup failed.');
        }

        try {
            // Install dnspython, junitparser, and sslyze
            const packageSetup: trm.ToolRunner = tl.tool(pyPath);
            packageSetup.arg('-m');
            packageSetup.arg('pip');
            packageSetup.arg('install');
            packageSetup.arg('-r');
            packageSetup.arg(path.join(__dirname, 'requirements.txt'));
            await packageSetup.exec();
            tl.setResult(tl.TaskResult.Succeeded, 'Python package install was successful.');
        
        } catch (err) {

            return tl.setResult(tl.TaskResult.Failed, 'Python package install failed.');
        }

        try {

            // Run the scan and generate the results
            const scan: trm.ToolRunner = tl.tool(pyPath);
            scan.arg(path.join(__dirname, './python/scanner.py'));
            scan.arg('--dns');
            scan.arg(`${dnsserver}`);
            scan.arg('--target');
            scan.arg(`${hostname}`);
            scan.arg('--port');
            scan.arg(`${port}`);
            scan.arg('--decision');
            scan.arg(`${testDecision}`);
            await scan.exec();
            
            return tl.setResult(tl.TaskResult.Succeeded, 'Scan ran successfully.');
        
        } catch (err) {
            tl.setResult(tl.TaskResult.Failed, 'Scan was unsuccessful.');

        }

    } catch (err: any) { // eslint-disable-line @typescript-eslint/no-explicit-any
        tl.error(err.message);
        tl.setResult(tl.TaskResult.Failed, tl.loc('taskFailed', err.message));

    }
}

run();


buildAndReleaseTask/getpythonpath.ts
/**
 * Python locater
 */

import { exist, which, tool as tool_1 } from 'azure-pipelines-task-lib/task';
import type trm from 'azure-pipelines-task-lib/toolrunner';
import * as tool from 'azure-pipelines-tool-lib/tool';
import * as path from 'path';

/* Location of python3 on Microsoft hosted agents:
https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/use-python-version?view=azure-devops */

export async function getPythonPath(): Promise<string> {

    if (exist('C:/hostedtoolcache/windows')) {
        // Windows
        console.log('AGENT: Extension running on a Microsoft hosted Agent.');
        const baseDir: string = tool.findLocalTool('python', '3.9');
        const pythonPath: string = path.join(baseDir, 'python.exe');

        return pythonPath;

    } else if (exist('/opt/hostedtoolcache')) {
        // Linux
        console.log('AGENT: Extension running on a Microsoft hosted Agent.');
        const baseDir: string = tool.findLocalTool('Python', '3.9');
        const pythonPath: string = path.join(baseDir, '/bin/python3');

        return pythonPath;

    } else if (exist('/Users/runner/hostedtoolcache')) {
        // OS X
        console.log('AGENT: Extension running on a Microsoft hosted Agent.');
        const baseDir: string = tool.findLocalTool('Python', '3.9');
        const pythonPath: string = path.join(baseDir, '/bin/python3');

        return pythonPath;

    } else {
        // Self-Hosted Agent
        console.log('AGENT: Extension running on a self-hosted Agent.');
        try {
            const pythonPath: string = await getSelfHostedPythonPath();

            return pythonPath;

        } catch (err: any) { // eslint-disable-line @typescript-eslint/no-explicit-any

            return err;
        }

    }
}

/**
 * Function to detect the Python path on a self-hosted ADO Agent
 */
async function getSelfHostedPythonPath(): Promise<string> {

    const selfHostedPythonPath: string = which('python3', true);

    if (selfHostedPythonPath != null) {
        const pythonVer: trm.ToolRunner = tool_1(selfHostedPythonPath);
        pythonVer.arg('-c');
        pythonVer.arg('import platform; print(platform.python_version())');
        // https://github.com/microsoft/azure-pipelines-task-lib/blob/master/node/docs/azure-pipelines-task-lib.md
        const result: string = await pythonVer.execSync().stdout;
        console.log('PYTHON VERSION: ' + `${result}`);

    } else {
        // Python3 not installed
        throw new Error('Python3 installation not found.');
    }

    return selfHostedPythonPath;

}


buildAndReleaseTask/python/scanner.py
""" Scanner module """

import argparse
import os

from dns import resolver
from junitparser import TestCase, TestSuite, JUnitXml, Failure
from sslyze.errors import ConnectionToServerFailed
from sslyze.plugins import scan_commands
from sslyze.plugins.openssl_cipher_suites.implementation import CipherSuitesScanResult
from sslyze.scanner.models import ServerScanRequest, ServerScanStatusEnum
from sslyze.scanner.scanner import Scanner
from sslyze.scanner.scan_command_attempt import ScanCommandAttemptStatusEnum
from sslyze.server_setting import ServerNetworkLocation


# SSl 2.0/3.0 and TLS 1.0/1.1 are prohibited cipher suites
CIPHER_SUITES = {
    scan_commands.ScanCommand.SSL_2_0_CIPHER_SUITES,
    scan_commands.ScanCommand.SSL_3_0_CIPHER_SUITES,
    scan_commands.ScanCommand.TLS_1_0_CIPHER_SUITES,
    scan_commands.ScanCommand.TLS_1_1_CIPHER_SUITES,
    scan_commands.ScanCommand.TLS_1_2_CIPHER_SUITES,
}

# Currently, only The following TLS 1.2 ciphers are considered "strong"
OK_TLS12_CIPHERS = {
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
    "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
}

# Create the parser
arg_parser = argparse.ArgumentParser(prog="scanner")

# Add the arguments
arg_parser.add_argument(
    "--target",
    "-t",
    metavar="target",
    type=str,
    required=True,
    help="The target to test",
)

arg_parser.add_argument(
    "--dns",
    "-d",
    metavar="dnsserver",
    type=str,
    required=True,
    help="The DNS server to use",
)

arg_parser.add_argument(
    "--port",
    "-p",
    metavar="port",
    type=int,
    required=False,
    default=443,
    help="The port to test (default is 443)",
)

arg_parser.add_argument(
    "--decision",
    metavar="decision",
    type=str,
    required=True,
    default="pass",
    choices=["fail", "pass"],
    help="Fail task if there are test failures?",
)

# Alias str type to better reflect the intented type and value
FQDN = str
IpAddr = str


def resolve_dnsname_to_ip(dns_server: IpAddr, dnsname: FQDN) -> IpAddr:
    """Resolve dns name and return IP"""
    ip_list = []  # results

    # dnspython config
    res = resolver.Resolver(configure=False)  # Do not read resolv.conf
    res.timeout = 3
    res.lifetime = 3  # How many seconds a query should run before timing out
    res.nameservers = [dns_server]  # DNS server may not be online
    #  Need better error handling in case internal DNS server is not online
    #  Right now it just times out the query, which is handled below

    try:
        answers = res.resolve(dnsname, search=False)
        for answer in answers.rrset:
            ip_list.append(answer.address)
        return ip_list[0]  # Return the first IP of the DNS Answer

    except resolver.NoAnswer as err:
        raise Exception(f"{dnsname} exists but no A record.") from err

    except resolver.NXDOMAIN as err:
        raise Exception(f"The DNS name {dnsname} does not exist.") from err

    except resolver.Timeout as err:
        raise Exception("The DNS operation timed out.") from err

    except resolver.NoNameservers as err:
        raise Exception(err.msg) from err  # Should trigger if a DNS server is offline

    except Exception as err:
        # If you are here, you are jacked
        raise ValueError("Catch all error in scanner.py") from err


def scan(dns_server: IpAddr, name: str, port: int) -> dict:
    """Three inputs: DNS server, web site name, and ip"""

    scan_output = new_results()
    scan_output["Hostname"] = name
    scan_output["DNS"] = dns_server

    # Validate hostname and resolve to IP
    try:
        target_ip = resolve_dnsname_to_ip(dns_server, name)
        scan_output["IP"] = target_ip

    except Exception as err:
        scan_output["Results"].append(err.args[0])
        return scan_output

    server_location = ServerNetworkLocation(name, port, target_ip)
    try:
        # pylint: disable=too-many-function-args
        scan_request = [
            ServerScanRequest(
                server_location=server_location, scan_commands=CIPHER_SUITES
            )
        ]
    except ConnectionToServerFailed as err:
        scan_output["Results"].append(err.args[0])
        return scan_output

    scanner = Scanner()
    scanner.queue_scans(scan_request)

    for results in scanner.get_results():
        if results.scan_status == ServerScanStatusEnum.ERROR_NO_CONNECTIVITY:
            scan_output["Results"].append(f"Error: Could not connect to {name}:{port}")
            return scan_output

        for suite in CIPHER_SUITES:
            tlsscanresult: CipherSuitesScanResult = getattr(
                results.scan_result, suite
            ).result
            tlsscanstatus = getattr(results.scan_result, suite).status
            if tlsscanresult.is_tls_version_supported:
                if tlsscanstatus == ScanCommandAttemptStatusEnum.ERROR:
                    scan_output["Results"].append(
                        {
                            "Version": f"{cipher_version}",
                            "Cipher": f"Error during scan of {cipher_version}",
                        }
                    )
                else:
                    cipher_version: str = tlsscanresult.tls_version_used.name
                    if cipher_version == "TLS_1_2":
                        for (
                            accepted_cipher_suite
                        ) in tlsscanresult.accepted_cipher_suites:
                            if (
                                accepted_cipher_suite.cipher_suite.name
                                not in OK_TLS12_CIPHERS
                            ):
                                scan_output["Results"].append(
                                    {
                                        "Version": f"{cipher_version}",
                                        "Cipher": f"{accepted_cipher_suite.cipher_suite.name}",
                                    }
                                )
                    else:
                        for (
                            accepted_cipher_suite
                        ) in tlsscanresult.accepted_cipher_suites:
                            scan_output["Results"].append(
                                {
                                    "Version": f"{cipher_version}",
                                    "Cipher": f"{accepted_cipher_suite.cipher_suite.name}",
                                }
                            )

    if len(scan_output["Results"]) == 0:
        scan_output["Results"].append("No SSL/TLS Violations found.")

    return scan_output


def new_results() -> dict:
    """Create the results dict"""

    return {"Hostname": None, "IP": None, "DNS": None, "Results": []}


def write_output(target, results) -> None:
    """Write scan results in junitxml format"""

    test_case = TestCase(f"{target}")
    test_case.name = f"{target}"
    if results["Results"] != ["No SSL/TLS Violations found."]:
        test_case.result = [Failure(results)]
    else:
        test_case.result = results

    suite = TestSuite("SSLChecker")
    suite.add_testcase(test_case)

    xml = JUnitXml()
    xml.add_testsuite(suite)
    xml.write("test-output.xml")


def main() -> None:

    """Main function"""
    # Execute the parse_args() method
    args = arg_parser.parse_args()
    target = args.target
    dns = args.dns
    arg_port = args.port
    if args.decision == "pass":
        decision = "false"
    else:
        decision = "true"

    scan_results = scan(dns, target, arg_port)

    write_output(target, scan_results)

    output = os.path.normpath(
        os.path.abspath(os.path.expanduser(os.path.expandvars("test-output.xml")))
    )

    # Borrowed from pytest-azurepipelines
    # https://github.com/tonybaloney/pytest-azurepipelines/blob/master/pytest_azurepipelines.py
    print(
        f"##vso[results.publish type=JUnit;runTitle='TlsTestGate';failTaskOnFailedTests={decision};publishRunAttachments=false;]{output}"  # pylint: disable=line-too-long
    )


if __name__ == "__main__":

    main()
